File: C++/Sorting/quick_sort.cpp
#include <iostream>
#define MAX 10
using namespace std;

void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
}

void quicksort(int [], int, int);

int main(){
    int Arr[MAX];
    int n;
    cin>>n;
    for(int i = 0; i < n - 1; i++)
        cin>>Arr[i];
    quicksort(Arr,0,n-1);
    cout<<"The sorted array is : ";
    for(int i = 0; i < n - 1; i++)
        cout<<Arr[i]<<" " ;
    return 0;
}

void quicksort(int arr[], int low, int high){
    if(low >= high)
        return;
    int mid = (low + high)/2;
    int left = low;
    int right = high;
    while(left <= right){
        while(arr[left] < arr[mid])//Move to misplaced element
            left++;
        while(arr[right] > arr[mid])//Move to misplaced element
            right--;

        //Swapping the misplaced element
        swap(&arr[left],&arr[right]);
        left++;
        right--;
    }

    quicksort(arr,low,mid);
    quicksort(arr,mid + 1,high);
}

File: Python/Sorting/quicksort.py
import random
num = int(input())
arr = [random.randint(0,100) for i in range(num)]

def quicksort(arr,low,high) : 
    part = partition(arr,low,high)
    print(part)
    quicksort(arr,low,part)
    quicksort(arr,part,high)

def partition(arr,low,high) : 
    left,right = 0,0
    print(arr[low:high])
    pivot = arr[low]
    while left <= right : 
        while arr[left] <= pivot and left < high-low : 
            left += 1
        while arr[right] >= pivot and right > low: 
            right -= 1
        arr[left],arr[right] = arr[right],arr[left]
        print(arr[low:high])
    return right
quicksort(arr,0,len(arr))
print(arr)
